### PR TITLE
Capability exercise cost

### LINKED ISSUE #1262: Automatically append command descriptions enabled by a device to that device's description
There is some redundancy in the documentation of a command and re-documentation in descriptions of devices that enable that command.

Instead, the description of a device, when displayed in the game UI, can be automatically augmented with a list of commands (and their descriptions) that are enabled by that device.

### LINKED ISSUE #1684: Entity consumption by commands
This proposal is to have a general mechanism to specify a cost, in the consumption of entities, for executing certain commands.

In the scenario `.yaml` file, an entity (specifically, a "device") may specify how many entities are consumed by the exercise of any of that device's capabilities (by way of executing a command).  Whereas currently capabilities are a list, they now becomes a map with the capability name as a key and an entity (to be consumed) as the value.  We may choose to implement `.yaml` parsing that retains backwards-compatibility of capabilities specified as a list.

*Aside: Alternatively, instead of a **single entity** as the map value, it could be an **ingredients list**.  However, for whatever command cost that consists of multiple ingredients, it would be equivalent to just define a recipe that outputs a single entity to be consumed, with the multiple ingredients as an input to that recipe.*

In the game, we first check whether a capability-providing device has the requisite entity inputs before it may invoke the command provided by that capability.  An entity shall be consumed regardless of the success/failure of the command itself.

If cost-free alternative devices exist to execute a given command, they are preferred.  Two devices that **both** have a cost are chosen based on their relative order in the scenario file.

## Examples
* The `ignite` command may be enabled by a `Zippo` device, which consumes `[1, lighter fluid]`.
* The `move` command may be enabled by a `rocket` device, which consumes `[1, propellant]`.
* The `detect` command may be enabled by a `blacklight` device, which consumes `[1, lumninol]`.

## Related
This may supersede/generalize #550, though "side-effects" like spilling water as a result of the `move` command would have to be specified on a per-effected-entity basis (as a map from `command` to `quantity lost`), as opposed to per-device as this proposal describes.

### PR DESCRIPTION
**Pull request by kostmo**
Closes #1684
Closes #1262

# Demo

A simple "puzzle" that makes use of consumables:

    scripts/play.sh -i data/scenarios/Testing/1777-capability-cost.yaml --autoplay

Demo of enabled commands and costs display in left pane:

    scripts/play.sh -i data/scenarios/Testing/1262-display-device-commands.yaml

Screenshot from 2024-03-01 22-39-12

# In this PR

* Supports specifying capabilities both as plain lists and as maps from capabilities to ingredients in YAML
* JSON Schema support for both capability specification styles
* New integration test
* Removed redundant `tshow` implementation from `Swarm.Doc.Util`

# Entity lookup approaches

The cost of exercising a capability in terms of "ingredients" is specified by the `_entityCapabilities` field within an `Entity` definition.  Each ingredient itself is an entity, specified by name.
For many purposes, having ingredients just of type `EntityName` is sufficient.  But for `Swarm.Game.Recipe.findLacking` in particular, the ingredients list must be actual `Entity` objects, not just names of entities.  So at some point the names need to be looked up in the global entity map to be promoted to `Entity` objects.

The full list of entities is not available at `Entity` parse time to look up an ingredient entity by name, so we cannot store ingredients lists of type `(Count, Entity)` within a parent `Entity` object.

Approaches considered were:

* Store a copy of the `entityMap` in `RobotR`, for use by the `equippedDevices` lens
* Introduce a type parameter to `Entity` representing a "parse phase"
* **Allow a redundant "entity lookup" (and validation) at command execution time**

## Store `entityMap` in `RobotR`

One approach explored was to add a field to `RobotR`:
```
  , _globalEntityMap :: EntityMap
```

This allowed the `equippedDevices` lens implementation to promote the `EntityName`s to `Entity`s when setting the value of the `_robotCapabilities` field.  However, it was rather invasive as it entailed threading the `EntityMap` through many new code paths.

## `Entity` type parameter

Currently, `Entity` has a field:
```
_entityCapabilities :: SingleEntityCapabilities EntityName
```

This would entail a huge refactoring, with:
```
data Entity e = Entity
  ...
  , _entityCapabilities :: SingleEntityCapabilities e
```

At initial parse time we would obtain a list of `Entity EntityName`, but somewhere later during `Scenario` parse time, we can do another pass to obtain `Entity Entity` objects.

This would at least have the advantage of doing the entity lookup/validation on ingredient lists in exactly one place, at parse time.

## Defer `EntityName -> Entity` promotion to command execution time

This is what is implemented in this PR.  The global set of capability costs is validated at scenario parse time.  But it is also redundantly validated in the `payExerciseCost` function, which is not ideal.

### CODE REVIEW

File: src/swarm-scenario/Swarm/Game/Device.hs [hunk 1]
```diff
@@ -0,0 +1,107 @@
...
+  deriving (Eq, Show, Generic, FromJSON, ToJSON, Hashable, Functor)
+
+instance (Eq e) => Ord (ExerciseCost e) where
+  compare = compare `on` (getCost . ingredients)
+
+data DeviceUseCost e en = DeviceUseCost
+  { device :: e
+  , useCost :: ExerciseCost en
+  }
+  deriving (Eq, Ord, Show, Generic, FromJSON, ToJSON, Functor)
```
[the remark this answers is not in the document]
Comment (byorgey) [reply src/swarm-scenario/Swarm/Game/Device.hs:89]: You can pass the `-ddump-deriv` flag to GHC to get it to show you the generated code for derived instances.  That would dump a LOT of code if we just compile all of Swarm with it, but maybe there's a way to get GHC to just compile a single file, I am not sure.

[the remark this answers is not in the document]
Comment (byorgey) [reply src/swarm-scenario/Swarm/Game/Device.hs:89]: In general the derived `Functor` instance will just map over the last type parameter, so the derived `fmap` here will do nothing to the `device` field and recursively call `fmap` on the `useCost` field, which would in turn call `fmap` on the `ingredients` field of `ExerciseCost`, and so on.

File: src/swarm-scenario/Swarm/Game/Device.hs [hunk 2]
```diff
@@ -0,0 +1,107 @@
...
+
+-- TODO Should this derive from an Applicative instance?
+promoteDeviceUseCost ::
+  Monad m =>
+  (e -> m e') ->
+  DeviceUseCost x e ->
+  m (DeviceUseCost x e')
+promoteDeviceUseCost f (DeviceUseCost d ex) =
+  DeviceUseCost d <$> transformIngredients f ex
```
[the remark this answers is not in the document]
Comment (byorgey) [reply src/swarm-scenario/Swarm/Game/Device.hs:98]: Off the top of my head, I'm pretty sure this should just be `traverse` from an automatically-derived `Traversable` instance for `DeviceUseCost x` (along with a similar `Traversable` instance for `ExerciseCost`).  In general there's only one sensible way to derive `Traversable` so we can be pretty confident it does the right thing.  If you're worried we can of course write some tests to exercise it and/or look at the output of `-ddump-deriv`.

File: src/swarm-scenario/Swarm/Game/Device.hs [hunk 3]
```diff
@@ -0,0 +1,107 @@
...
+
+-- TODO Should this derive from an Applicative instance?
+transformIngredients ::
+  Monad m =>
+  (e -> m e') ->
+  ExerciseCost e ->
+  m (ExerciseCost e')
+transformIngredients f (ExerciseCost ings) =
+  ExerciseCost <$> mapM (traverse f) ings
```
[the remark this answers is not in the document]
Comment (byorgey) [reply src/swarm-scenario/Swarm/Game/Device.hs:107]: Yes, note that the `Monad m` constraint is unnecessary, we can get away with `Applicative m`, if we rewrite `mapM (traverse f) ings` as `traverse (traverse f) ings`, or more idiomatically, `(traverse . traverse) f ings`.  But that's exactly what the derived instance of `Traversable` will do.

File: src/swarm-scenario/Swarm/Game/Device.hs (hunk 2 above)
[the remark this answers is not in the document]
Comment (kostmo) [reply src/swarm-scenario/Swarm/Game/Device.hs]: Ok, was able to used the derived `Traversable`.

File: src/Swarm/TUI/View.hs [hunk 4]
```diff
@@ -1236,6 +1240,66 @@ displayProperties = displayList . mapMaybe showProperty
...
+explainCapabilities :: GameState -> Entity -> Widget Name
+explainCapabilities gs e
+  | null capabilitiesAndCommands = emptyWidget
+  | otherwise =
+      padBottom (Pad 1) $
+        vBox
+          [ hBorderWithLabel (txt "Enabled commands")
+          , hCenter
+              . vBox
+              . L.intersperse (padTop (Pad 1) . hCenter . txt $ T.replicate 10 "*")
```
Comment (byorgey) [inline src/Swarm/TUI/View.hs:1256]: I'm not a big fan of the asterisk dividers; they don't seem to fit the rest of the visual aesthetic and they waste vertical space.  Do we need dividers between the command descriptions at all?  I think just a blank line of space between commands ought to be fine.

File: src/Swarm/TUI/View.hs [hunk 5]
```diff
@@ -1236,6 +1240,66 @@ displayProperties = displayList . mapMaybe showProperty
...
+      , padLeft (Pad 1) . vBox . map renderCmdInfo . NE.toList $ enabledCommands cmdsAndCost
+      ]
+
+  renderCmdInfo c =
+    padTop (Pad 1) $
+      vBox
+        [ hBox
+            [ padRight (Pad 1) (txt . syntax $ constInfo c)
+            , padRight (Pad 1) (txt ":")
+            , withAttr magentaAttr . txt . prettyText $ inferConst c
```
Comment (byorgey) [inline src/Swarm/TUI/View.hs:1283]: Many of the types are getting pretty-printed on multiple lines when there is in fact plenty of horizontal space to print them on one line.  I looked into what `prettyText` does, and it eventually calls `layoutPretty defaultLayoutOptions`.  We should look into making a variant of `prettyText` that takes the available space of the current brick rendering context into account.

File: src/swarm-scenario/Swarm/Game/Device.hs [hunk 6]
```diff
@@ -0,0 +1,108 @@
...
+  { getMap :: Map Capability e
+  }
+  deriving (Show, Eq, Generic, ToJSON, Hashable, Functor, Foldable, Traversable)
+
+getCapabilitySet :: Capabilities e -> Set Capability
+getCapabilitySet (Capabilities m) = M.keysSet m
+
+type SingleEntityCapabilities e = Capabilities (ExerciseCost e)
+
+type MultiEntityCapabilities e en = Capabilities (NonEmpty (DeviceUseCost e en))
```
[the remark this answers is not in the document]
Comment (kostmo) [reply src/swarm-scenario/Swarm/Game/Device.hs:51]: Added a comment (to the `DeviceUseCost` type).

File: src/swarm-scenario/Swarm/Game/Device.hs [hunk 7]
```diff
@@ -0,0 +1,108 @@
+-- |
+-- SPDX-License-Identifier: BSD-3-Clause
+--
+-- A device is an entity that provides capabilities.
+--
+-- Some capabilities have a cost to exercise.
```
[the remark this answers is not in the document]
Comment (kostmo) [reply src/swarm-scenario/Swarm/Game/Device.hs:6]: The suggested comments are good.

File: src/Swarm/TUI/View.hs (hunk 5 above)
Comment (kostmo) [reply src/Swarm/TUI/View.hs:1283]: #1794

File: src/Swarm/TUI/View.hs (hunk 4 above)
Comment (kostmo) [reply src/Swarm/TUI/View.hs:1256]: Fixed.  Sections are now delimited by two blank lines.   I think indentations do a sufficient job delimiting the other content.

Screenshot from 2024-04-24 22-54-18

File: src/swarm-scenario/Swarm/Game/Device.hs [hunk 8]
```diff
@@ -0,0 +1,108 @@
...
+  }
+  deriving (Generic, FromJSON)
+
+-- | First, attempt to parse capabilities as a list.
+-- Otherwise, parse as a Map from capabilities to ingredients.
+instance (FromJSON e) => FromJSON (SingleEntityCapabilities e) where
+  parseJSON x =
+    Capabilities <$> (simpleList <|> costMap)
+   where
+    simpleList = M.fromSet (const $ ExerciseCost []) <$> parseJSON x
```
[the remark this answers is not in the document]
Comment (kostmo) [reply src/swarm-scenario/Swarm/Game/Device.hs]: Fixed

**Comment by kostmo** [comment]
All review comments now addressed, @byorgey.

**Review by byorgey — APPROVED** [review]
:+1:

**Pull request merged**